### PR TITLE
Handle huge unions better in createUnionOrIntersectionProperty

### LINKED ISSUE #29834: Opening/editing project file causes 100% CPU usage by tsserver
- VSCode Version: 1.31.0-insider (1.31.0-insider) (also tested in latest non-insider release)
- OS Version: macOS 10.14.2 (18C54)

**Steps to Reproduce:**
1. Clone this repo: https://github.com/pqvst/vscode-cpu-issue-test
2. Open the folder in VSCode
3. Open `browse.js`
4. tsserver CPU usage goes to 100% for a while
5. Reopening the file or making any edit will again trigger high CPU usage

**Expected results:**
I expect to be able to work in my code base without having my fans go crazy.

**Actual results:**
Working in my code base causes my fans to go crazy.

**Does this issue occur when all extensions are disabled?:**
Yes

screenshot 2019-02-06 13 16 18

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #29834 - or at least makes it more tolerable.

Timeline before this fix, roughly 7 seconds until first results for a quickinfo relying on the 1MB data object:
image

Timeline after this fix for the same info:
image
Down to 2s. (Specifically, time in `createUnionOrIntersectionProperty ` was cut from 5s down to 340ms)

That's a pretty big improvement for just fixing property creation so it's not O(n^2) with property count. 😄 

There's not really a test associated with this change because, well, perf...

### CODE REVIEW

**Comment by weswigham** [comment]
There's not really a noticeable difference in check time one way or the other in the `Compiler - Unions` perf test.

image

**Comment by weswigham** [comment]
@typescript-bot perf test 🕙

**Comment by weswigham** [comment]
@sandersn There's no perf diff between [this branch](https://typescript.visualstudio.com/TypeScript/_build/results?buildId=25487) and [`master`](https://typescript.visualstudio.com/TypeScript/_build/results?buildId=25465).

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -7809,7 +7812,8 @@ namespace ts {
                     }
                 }
             }
-            if (!props) {
+            const props = arrayFrom(propSet.values());
```
Comment (sheetalkamat) [inline src/compiler/checker.ts]: Why not check propSet.size to avoid creating emptyArray to just check size?

Comment (sheetalkamat) [inline src/compiler/checker.ts]: Can we guarantee order of properties created from map? We seem to rely on the order eg to set `firstValueDeclaration`

Comment (weswigham) [reply src/compiler/checker.ts]: Yes. More reliably then objects, even. From [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map):

> The keys in Map are ordered while keys added to object are not. Thus, when iterating over it, a Map object returns keys in order of insertion.

Comment (weswigham) [reply src/compiler/checker.ts]: Sure, why not. Done.

**Pull request merged**